modules/kargoBidAdapter.js

```diff
@@ -28,57 +28,96 @@ export const spec = {
 
-    const transformedParams = Object.assign({}, {
-      sessionId: spec._getSessionId(),
-      requestCount: spec._getRequestCount(),
+    const krakenParams = Object.assign({}, {
+      pbv: '$prebid.version$',
+      aid: firstBidRequest.auctionId,
+      sid: spec._getSessionId(),
+      url: spec._getAllMetadata(bidderRequest).pageURL,
       timeout: bidderRequest.timeout,
-      currency,
-      cpmGranularity: 1,
-      timestamp: (new Date()).getTime(),
-      cpmRange: {
-        floor: 0,
-        ceil: 20
-      },
-      bidIDs,
-      bidSizes,
+      ts: new Date().getTime(),
       device: {
-        width: window.screen.width,
-        height: window.screen.height
+        size: [
+          window.screen.width,
+          window.screen.height
+        ]
       },
-      prebidRawBidRequests: validBidRequests
-    }, spec._getAllMetadata(bidderRequest, tdid));
+      imp: impressions,
+      user: spec._getUserIds(tdid, bidderRequest.uspConsent, bidderRequest.gdprConsent),
+      eids: firstBidRequest.userIdAsEids
+    });
+
+    const reqCount = spec._getRequestCount()
+    if (reqCount != null && reqCount > 0) {
+      krakenParams.requestCount = reqCount
+    }
+
+    if (currency != null && currency != 'USD') {
+      krakenParams.cur = currency
+    }
+
+    // Pull Social Canvas segments and embed URL
+    const socialCanvas = deepAccess(firstBidRequest, 'params.socialCanvas');
+
+    if (socialCanvas != null) {
+      krakenParams.socan = socialCanvas
+    }
 
     // User Agent Client Hints / SUA
     const uaClientHints = deepAccess(firstBidRequest, 'ortb2.device.sua');
     if (uaClientHints) {
-      transformedParams.device.sua = pick(uaClientHints, ['browsers', 'platform', 'mobile', 'model']);
+      const suaAttributes = ['browsers', 'platform', 'mobile', 'model', 'source']
+      let suaValidAttributes = []
+
+      suaAttributes.forEach(attributeKey => {
+        var attributeValue = uaClientHints[attributeKey]
+        if (attributeValue != null) {
+          if (((attributeKey == 'mobile' || attributeKey == 'source') && attributeValue < 1) ||
+            (attributeKey == 'model' && attributeValue.trim() == '')) {
+            return
+          }
+          suaValidAttributes.push(attributeKey)
+        }
+      })
+
+      krakenParams.device.sua = pick(uaClientHints, suaValidAttributes);
     }
 
-    // Pull Social Canvas segments and embed URL
-    const socialCanvas = deepAccess(firstBidRequest, 'params.socialCanvas');
-    if (socialCanvas) {
-      transformedParams.socialCanvasSegments = socialCanvas.segments;
-      transformedParams.socialEmbedURL = socialCanvas.embedURL;
+    const validPageId = spec._getLocalStorageSafely('pageViewId') != null 
+    const validPageTimestamp = spec._getLocalStorageSafely('pageViewTimestamp') != null
+    const validPageUrl = spec._getLocalStorageSafely('pageViewUrl') != null
+ 
+    if(validPageId || validPageTimestamp || validPageUrl) {
```

Comment:
can consider splitting this up for better readability:
```
    const page = {}
    if (validPageId) {
      page.id = spec._getLocalStorageSafely('pageViewId')
    }
    if(validPageTimestamp) {
      page.timestamp = spec._getLocalStorageSafely('pageViewTimestamp')
    }
    if(validPageUrl) {
      page.url = spec._getLocalStorageSafely('pageViewUrl')
    }
    if (_.isEmpty(page) !== '{}') {
      krakenParams.page = page
    }
``

Reply:
yep, looking at this again it wasnt clear. Changed